### PR TITLE
Solids - enable use of GPU vectors

### PR DESCRIPTION
**Pull request by jeremylt**
Once #564 is merged, I will update and test this PR.

### CODE REVIEW

File: examples/solids/src/matops.c [hunk 1]
```diff
@@ -34,19 +34,20 @@ PetscErrorCode ApplyLocalCeedOp(Vec X, Vec Y, UserMult user) {
...
   CeedOperatorApply(user->op, user->Xceed, user->Yceed, CEED_REQUEST_IMMEDIATE);
-  CeedVectorSyncArray(user->Yceed, CEED_MEM_HOST);
+  CeedVectorSyncArray(user->Yceed, user->memType);
 
   // Restore PETSc vectors
-  ierr = VecRestoreArrayRead(user->Xloc, (const PetscScalar **)&x);
+  ierr = user->VecRestoreArrayRead(user->Xloc, (const PetscScalar **)&x);
   CHKERRQ(ierr);
-  ierr = VecRestoreArray(user->Yloc, &y); CHKERRQ(ierr);
+  ierr = user->VecRestoreArray(user->Yloc, &y); CHKERRQ(ierr);
```
Comment (jeremylt) [inline examples/solids/src/matops.c:50]: Debugging note - something in these 14 lines is breaking our multi-component PETSc examples.

Comment (jeremylt) [reply examples/solids/src/matops.c:50]: If I transfer host pointers from PETSc to libCEED, then the code works correctly. If I pass device pointers, then the code does not. We have the same issue with `petsc-bps`.

Comment (jeremylt) [reply examples/solids/src/matops.c:50]: Note - this problem doesn't present for single component PETSc examples, like BP1 or BP3.

Comment (jeremylt) [reply examples/solids/src/matops.c:50]: Solids works, still tracking the bp/multigrid bug

**Comment by jeremylt** [comment]
Got the bug (or more accurately, a whole slew of bugs).

Ready for review.

File: examples/petsc/bpsraw.c [hunk 2]
```diff
@@ -878,7 +878,7 @@ int main(int argc, char **argv) {
     ierr = KSPSetTolerances(ksp, 1e-10, PETSC_DEFAULT, PETSC_DEFAULT,
                             PETSC_DEFAULT); CHKERRQ(ierr);
   }
-  ierr = KSPSetOperators(ksp, mat, mat); CHKERRQ(ierr);
+  ierr = KSPSetOperators(ksp, mat, mat);
```
Comment (jedbrown) [inline examples/petsc/bpsraw.c:881]: Why?

Comment (jeremylt) [reply examples/petsc/bpsraw.c:881]: woops

Comment (jeremylt) [reply examples/petsc/bpsraw.c]: fixed

File: examples/petsc/bpsraw.c [hunk 3]
```diff
@@ -232,7 +232,7 @@ bpData bpOptions[6] = {
     .error = Error3,
     .setupgeofname = SetupDiffGeo_loc,
     .setuprhsfname = SetupDiffRhs3_loc,
-    .applyfname = Diff_loc,
+    .applyfname = Diff3_loc,
```
Comment (valeriabarra) [inline examples/petsc/bpsraw.c:235]: Oh wow, thanks for spotting this!

Comment (jeremylt) [reply examples/petsc/bpsraw.c:235]: It might be the cause of #451. Let me check.

Comment (jeremylt) [reply examples/petsc/bpsraw.c:235]: Nope

Comment (valeriabarra) [reply examples/petsc/bpsraw.c:235]: :/

**Comment by jeremylt** [comment]
Grr, found a different bug, squashed it, and updated test parameters to help catch it.

**Comment by jeremylt** [comment]
I thought about ditching the `VecGetArray` ect, but I ended up retaining them because they were useful for debugging purposes.

**Comment by jeremylt** [comment]
I can update the manpage, but I think it should say that users either use [ `GetArray` and `RestoreArray`] or [`SetArray` and `SyncArray`].

**Comment by jedbrown** [comment]
Yeah, perhaps explain that in the manual. Also that it's a no-op if the preferred memtype of your backend matches the memtype that you've Set. So if one always matches CUDA backends with PETSc CUDA Vecs and host with host, then the call could be dropped. (We can recommend not doing that because it's useful for debugging.)

Should we add an interface to PETSc where the desired memory space is an explicit argument, instead of implicit as in `VecGetArrayInPlace`?

**Comment by jeremylt** [comment]
Updated

**Comment by jeremylt** [comment]
We perhaps want to consider a `CeedVectorResetArray(CeedVector vec)` at some point (which would just be a clearer way to call `CeedVectorSetArary(vec, CEED_MEM_HOST, NULL)` to detach an external pointer from a libCEED vector.

**Comment by jedbrown** [comment]
`CeedVectorTakeArray` to reclaim it? I agree it's dirty and error-prone to leave the stale pointer in there.

**Comment by jeremylt** [comment]
I was thinking of basically: https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecResetArray.html

We could also do a version that does Sync + Reset.

I'll make an issue

**Pull request merged**